Get the gitignore cache working, consistently use fast JSON modules

gsnedders opened the pull request:
Some Sam Sneddon person broke the gitignore cache with e98eee5aec6 (as MutableMapping provides a default `__bool__`, effectively returning `bool(len(self))`), and this gets it working again.

This also slightly more consistently gets us using quicker JSON modules, and factors out a bunch of code for this which uses whichever we want. This provides a further benefit for those with them installed (`ujson` especially).

A few benchmarks (Py3.8, with a modified untracked file in the tree):

* master: 3.01s
* this branch: 2.82s
* this branch + orjson: 2.80s
* this branch + ujson: 2.17s
* this branch + orjson + ujson: 2.16s
* this branch + orjson + ujson patched for https://github.com/ultrajson/ultrajson/issues/429: 1.99s

(I don't remember why we set `caller_owns_obj=False` in `manifest.write`, given it exists for that reason, but with that set to `True` and with all the above we're down to 1.52s.)

gsnedders committed:
Avoid relying on curdir being "" and not "."

At the moment it's always ".", but it's a non-obvious performance
trap if we ever do change the behaviour here

gsnedders committed:
Get the gitignore cache actually working

gsnedders committed:
Use quicker JSON modules as much as possible

Hexcles commented:
Curious why we adding yet another fast JSON module fallback (orjson)?

gsnedders commented:
> Curious why we adding yet another fast JSON module fallback (orjson)?

When we have it shared in a common place, the cost of adding it is pretty low. That said, for all that's been posted in many places about its performance its gains are… pretty minute. Happy to remove it, given that.

jgraham commented on tools/manifest/fastjson.py:
```diff
@@ -0,0 +1,162 @@
+
+
+#
+# dump/dumps_dist (for distributed usage of JSON where files should safely roundtrip)
```
I think I'd use "stable" rather than "dist". And maybe just make it a kwarg.

gsnedders replied:
"stable" to me implies merely that it's idempotent, rather than it produces the same result in all configurations (which is what we aim for for the MANIFEST as far as I'm aware).

jgraham replied:
OK, well I don't really mind either way.

jgraham commented on tools/manifest/vcs.py:
```diff
@@ -237,27 +241,35 @@ def check_valid(self, data):
     def __contains__(self, key):
         # type: (Any) -> bool
+        try:
+            key = key.decode("utf-8")
```
Ugh, this is depressing. Can we avoid all the decode/encode stuff somehow?

gsnedders replied:
The decision seemingly was made to use bytes for file names around gitignore while I was away; I was always skeptical about that option because of the implications on the cache (having to decode/encode everywhere). The alternative is that we use unicode for filenames, then we don't need to encode/decode it for JSON.

jgraham replied:
Yeah, I guess this is the thing where we want to handle all possible paths in a rather generic module, but then using JSON as the serialization format doesn't work. I suppose we could instead write the cache as a null-separated flat file of `path\0value\0` or similar.

gsnedders commented on tools/manifest/fastjson.py:
```diff
@@ -0,0 +1,162 @@
+}  # type: Dict[str, Any]
+
+if PY3:
+    __ujson_dump_dist_kwargs['reject_bytes'] = True
```
Huh? I don't see how 3.5 is relevant here? `ujson` by default allows bytes or unicode for strings on both Python 2 and 3; having the behaviour differ with the stdlib on Python 3 seems harmful and liable to end up with bugs being introduced, so this option makes behaviour match the stdlib.

jgraham replied:
Maybe I'm thinking of parsing bytes then.

gsnedders commented on tools/manifest/manifest.py:
```diff
@@ -13,6 +12,7 @@
     string_types,
 )
 
+from . import fastjson
```
What would you rather call it? I called it `fast` because that's what you called the import in manifest!

jgraham replied:
Silly me ;p Maybe just jsonlib or something.

gsnedders commented on tools/manifest/fastjson.py:
```diff
@@ -0,0 +1,162 @@
+#
+# dump/dumps_local options for some libraries
+#
+__ujson_dump_local_kwargs = {
```
Is a `_` name and `__all__` enough to make you happy?

jgraham replied:
Yes.

gsnedders committed:
Rename the gitignore.json cache file

Problematically, older versions of the manifest on Python 3 break after
this PR when the gitignore.json cache file exists.

gsnedders committed:
Remove the shortlived support for orjson

gsnedders committed:
fixup! Remove the shortlived support for orjson

gsnedders committed:
Avoid __* names

gsnedders committed:
jgraham regrets the fastjson name
